python/lib/core/dmod/core/common/collections/bag.py

```diff
@@ -0,0 +1,126 @@
+"""
+Defines a collection of unordered data not requiring a hash
+"""
+from __future__ import annotations
+
+import typing
+
+from typing_extensions import Self
+
+
+_T = typing.TypeVar("_T")
+
+
+class Bag(typing.Collection[_T]):
```

Comment:
Items in a `collections.Counter` _must_ be hashable. If that constraint could be removed then we wouldn't need this class. 

Bisect is pretty cool, but this isn't meant to be a high speed collection. I'd be shocked if this would ever hold anything over the 1,000,000 or so items it would require to make the speedier operations necessary considering how incredibly niche the use cases are.

Reply:
The annotation isn't marked as orderable since there isn't a vanilla orderable generic type or protocol defined. I agree it would be nice, though.

Reply:
> Items in a `collections.Counter` _must_ be hashable. If that constraint could be removed then we wouldn't need this class.

Right, I guess my point is this collection feels more and more like strictly a convenience. You could make any type unhashable `T` hashable by wrapping it in a generic hashable wrapper type.

Reply:
Yeah, for sure! Sorry for being vague!

```python
import abc
import dataclasses
import functools

T = typing.TypeVar("T")

# functional inspired version
@dataclasses.dataclass
class Hash(typing.Generic[T]):
    value: T
    hasher: typing.Callable[[T], int] = dataclasses.field(repr=False)

    def __hash__(self) -> int:
        return self.hasher(self.value)

    @classmethod
    def bind_hasher(cls, hasher: typing.Callable[[T], int]):
        return functools.partial(cls, hasher=hasher)

# classic OOP version
class HashABC(typing.Generic[T], abc.ABC):
    def __init__(self, value: T):
        self.value = value

    @abc.abstractmethod
    def __hash__(self) -> int:
        ...
```

You can imagine building class specific hashers that for example, use fields in a more complex object to create a unique hash. Of course you could just implement `__hash__` on the original object, but that might not be desirable nor possible. It would then be trivial to create generic container versions of a hasher that for example, given a `Iterable[Hasher]` computes all the inner hashes and sums them and returns that as the aggregate hash. This would let you wrap an unhashable type in a hashable wrapper and then use the wrapper in a collection that requires `__hash__`.